### DISCUSSION TITLE
Support for NER models

### DISCUSSION THREAD

**Discussion by pabvald** [cdpierse/transformers-interpret#83] [Q&A] [answered]
Hi! 

I have tried to generate explanations for NER models but I haven't found any description of how to implement this, so I have tried to extrapolate what is been already done for other tasks, such as multi-class classification. 

I have followed the approach described in the [Captum tutorial to interpret BERT](https://captum.ai/tutorials/Bert_SQUAD_Interpret) in which a custom forward function that allows to access a particular position of the prediction using the `position` input argument is used and the attributions are computed with respect to the `BertEmbeddings` using the `captum.attr.LayerIntegratedGradients` class. 

I have generated two types of visualizations: 

1. Attributions considering each token (position) and its corresponding predicted NER label: this allows to see how the rest of the sentence influenced each token to be classified into a particular NER class.
2. Attributions considering a single token and all the possible NER classes: this allows to see how the rest of the sentence influenced a certain token to be classified into a class and not into any of the others.

A notebook with these two visualizations can be found in [this repo](https://github.com/pabvald/ner-explainability). I would like to know if my approach makes sense, or, if not,  what I should do to correctly visualize explanations for NER models. Once I am sure my approach is right, I would like to contribute to **Transformers Interpret** implementing the support for NER models.

Thank you in advance for the help!

**Comment by cdpierse** [comment, marked as answer]
Hi @pabvald,

Thank you very much for putting the time into looking into this. I'd love to get the NER explainer finished to round out the capabilities of the text part of the package. 

I think the approach you are taking for the first visualization is the right way to go. For most people my assumption is that they will want to just explore why a certain word was given a NER label, if there are requests for exploring all possible classes we could explore that at a later time, but the first use case seems like the most practical. 

Your approach seems to be right from what I've looked at so far, would be delighted if you were up for contributing this feature, it'd be a major boost for the package. If you take a look at this [Multi-Label Explainer
](https://github.com/cdpierse/transformers-interpret/blob/master/transformers_interpret/explainers/multilabel_classification.py) it should give a good idea of how you can use the existing explainer class and utils to create a subclassed NER explainer. One custom feature I wanted to include for NER was the ability to pass the name or index of certain NER labels that are to be ignored by the explainer. This will make explanations for very long sentences much nicer to read. 

If you need any help at all with this let me know.

Thanks again, 
Charles

**Reply by pabvald** [reply]
Hi @cdpierse,

great! I will work on it and include the feature you suggest. 

Thank you for you answer, 
Pablo
